<a href="https://colab.research.google.com/github/pedroamericovinhas/math/blob/main/calculus/Untitled0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Given a function $P(x)$, we can approximate its [Riemann Integral](https://en.wikipedia.org/wiki/Riemann_integral) using python.

In [1]:
def P(x):
  return -0.5*x*x + 10

A first rough aproximation could be defined by summing integer values inside the interval of $f$ we want to integrate.

From the [wikipedia article on the Euler-Maclaurin formula](https://https://en.wikipedia.org/wiki/Euler%E2%80%93Maclaurin_formula):
>If $m$ and $n$ are natural numbers and $f(x)$ is a real or complex valued continuous function for real numbers $x$ in the interval $[m,n]$, then the integral
> $$I=\int^n_mf(x)\;dx$$
> can be approximated by the sum
> $$S = f(m+1) + \cdots + f(n-1) + f(n)$$

Let's code this!

In [2]:
def riemann_sum(f,m,n):
  # m, n ∈ N
  res = sum(f(x+1) for x in range(m,n))
  # f(m+1) + ⋯ + f(n-1) + f(n)
  return res

riemann_sum(P,0,4) # Actual value of integral: 29.333…

25.0

As you could've probably guessed, it's a pretty bad approximation; though it hints at how we might proceed: reducing the step size.

`range()` doesn't accept non-integer step size, so we'll have to create a new method to do this. Let's say we wanted $0.1$ to be our step size. We could calculate

$f(a+0.1) + f(a+0.2) + \cdots + f(b-0.1) + f(b)$

And get the exact result we want! 

...That is, if $\; b-a \;$ is a clean multiple of $0.1$. Say our bounds are $[0.2,\; 0.45]$. In this case, we run into a problem:

$f(0.3) + f(0.4) \; + \;?$

Suddenly, we've lost 20% of our interval. However, this lost interval tends to decrease as we decrease our step size. If it was $0.01$ or $0.001$, we would only notice a difference in very small intervals. In fact, recall that an *actual* integral would have this step size approaching $0$, so the last dropped term would be completely negligible. One way to code this is:
>```python
(f(a + x*dx) for x in range(0, (b-a) // dx))
>```

where $dx$ stands for our (very small) step size.

Let's define this $dx$ variable globally:

In [3]:
dx = 0.01

At the end of our summing process, we'll have to multiply the result by $dx$. This is in order to compensate for the smaller step size, and can be geometrically explained in the following manner:

> Our process involves summing multiple rectangles under the function $f$. These rectangles have height $f(x)$ and width $dx$ (*that's how we defined $dx$!*). 
>
>>$Area = h_1⋅dx + h_2⋅dx + \cdots + h_{n-1}⋅dx + h_n⋅dx $
> 
> Where $h_i$ are the different heights of the rectangles.
>
> Using the distributive property of multiplication, we can pull those $dx$ terms from out of the sum
>
>>$Area = dx(h_1 + h_2 + \cdots + h_{n-1} + h_n)$ 

Of course, we implicitly did this in our first approximation, where $\; dx = 1$.

Putting it all together, we get:

In [4]:
dx = 0.01
def integral(f,a,b):
  res = sum(f(a+x*dx) for x in range(0, int((b-a)//dx)))
  return res * dx
integral(P, 0, 4)

29.3529005

That's much better! And it only gets better for smaller values of $dx$. For comparison's sake, here's a table with various values of $dx$:

|$dx$|Result
|---|---|
|0.1|29.490499999999987|
|0.05|29.422562499999998|
|0.01|29.3529005|
|0.005|29.343225062499997|
|0.001|29.335329000500092|
|0.0001|29.333533290000368|
|0.0000001|29.3333337333407|


---


That's cool and all, but let's do something even cooler with integrals: estimating the value of π!

You may or may not be familiar with this formula:

$$\pi = 4\int_0^1\frac{1}{1+x^2}\; dx$$

We can very easily calculate this using our integral function, with the help of python's `lambda` functions. We can also afford to use much smaller values of $dx$, given our relatively small integration bounds.

In [5]:
dx = 0.0000001
pi = 4 * integral(lambda x : 1/(1+x*x), 0, 1)
pi

3.141592753589987

That's only 0.000003% bigger than π.